### ISSUE TITLE
[STORM-2914] Remove enable.auto.commit support from storm-kafka-client

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#6696]
<p>The enable.auto.commit option causes the KafkaConsumer to periodically commit the latest offsets it has returned from poll(). It is convenient for use cases where messages are polled from Kafka and processed synchronously, in a loop. </p>

<p>Due to <a href="https://issues.apache.org/jira/browse/STORM-2913" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-2913</a> we'd really like to store some metadata in Kafka when the spout commits. This is not possible with enable.auto.commit. I took at look at what that setting actually does, and it just causes the KafkaConsumer to call commitAsync during poll (and during a few other operations, e.g. close and assign) with some interval. </p>

<p>Ideally I'd like to get rid of ProcessingGuarantee.NONE, since I think ProcessingGuarantee.AT_MOST_ONCE covers the same use cases, and is likely almost as fast. The primary difference between them is that AT_MOST_ONCE commits synchronously.</p>

<p>If we really want to keep ProcessingGuarantee.NONE, I think we should make our ProcessingGuarantee.NONE setting cause the spout to call commitAsync after poll, and never use the enable.auto.commit option. This allows us to include metadata in the commit.</p>

---
<details><summary><i>Originally reported by <a title="Stig Rohde Døssing" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-2914" target="_blank">Remove enable.auto.commit support from storm-kafka-client</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Stig Rohde Døssing" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-02-06T04:51:12+00:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p>Ping <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a> and <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a>, I'd appreciate your opinions on this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen">avermeerbergen</a>:</i>
<p>Hello <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a>,</p>

<p>We use auto-commit when consuming Kafka messages in our "real-time alerting" topology.</p>

<p>This topology doesn't uses acking nor anchoring</p>

<p>Indeed, the purpose of our "real-time alerting" topology is to evaluate metrics read from Kafka through expressions (called triggers, like in Zabbix) into "alerting severities", which we write to both a Redis database (for rendering in an alerting web app) and into other Kafka topics (for notifications / persistency purposes handled by other topologies).</p>

<p>We observed that sometimes our alerting topology is flooded by an excessive rate of metrics = Kafka messages read by kafka spout, because sometimes a remote Kafka cluster which was temporarily unable to replicate data to our central Kafka cluster (the one plugged to our topologies) comes "back to life" and sends all metrics at once, including metrics which are too old to be relevant for real-time alerting.</p>

<p>And because if Storm 1.x back-pressure is kind of ... well, say "limited", the result of such flooding is that our topology is just stuck and no longer consumes any metric.</p>

<p>So we have a simple self-healer "cron" which periodically checks whether or not our topology consumes metrics over a sliding windows of 10 minutes, and when consumption is stopped for 10 minutes, this cron will restart our topology with "LATEST " strategy. </p>

<p> </p>

<p>No with this context, after trying many combinations we found out that with Storm 1.1.0 and with our own Kafka spout (working in auto-commit mode, but not as powerful as storm kafka client), we had to enable auto-commit otherwise our topology would get stuck quite too often.</p>

<p>We had been waiting for Storm 1.2.0 to use the official storm-kafka-client  with auto-commit support, because our own spout was quite limited.</p>

<p>Now, in order to answer your question about storm-kafka-client removal of auto-commit support, I am unsure about what I should answer: what would guarantee me that my real-time alerting topology's stability won't be worse than it was when I used storm-kafka-client before the following commit broke it :<a href="https://github.com/apache/storm/commit/a3899b75a79781602fa58b90de6c8aa784af5332#diff-7d7cbc8f5444fa7ada7962033fc31c5e" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/commit/a3899b75a79781602fa58b90de6c8aa784af5332#diff-7d7cbc8f5444fa7ada7962033fc31c5e </a>?</p>

<p>Best regards,</p>

<p>Alexandre Vermeerbergen</p>

<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a></p>

<p>Regarding your use case, you might try at-least-once mode plus a filter to discard old metrics, plus the topology.max.spout.pending parameter to limit how many messages go into your topology at a time. That said, I'll continue on with the assumption that your current configuration is what you want to keep.</p>

<p>I think it's likely that use cases like yours where you don't need at-least-once would be equally well supported by using the AT_MOST_ONCE processing guarantee <a href="https://github.com/apache/storm/blob/master/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java#L165" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/master/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java#L165</a>. </p>

<p>The main difference between enable.auto.commit and AT_MOST_ONCE is that AT_MOST_ONCE synchronously commits offsets to Kafka once it has polled, but before the tuples are emitted, while enable.auto.commit commits every once in a while before the consumer polls. What this means is that enable.auto.commit may emit a tuple 0, 1, 2, etc. times, while AT_MOST_ONCE will emit a tuple 0 or 1 times. Since users of enable.offset.commit don't care how many times a message is processed, they should be fine with the AT_MOST_ONCE guarantee.</p>

<p>My only concern with dropping support for enable.auto.commit and asking people to use AT_MOST_ONCE instead, would be the switch to synchronous committing, which may have some impact on performance. I would be surprised if the impact were meaningful though. </p>

<p>For now I think we should try to make the spout emulate enable.auto.commit when the processing guarantee is NONE, without actually setting enable.auto.commit on the consumer. That way we can specify metadata when committing, which allows us to fix <a href="https://issues.apache.org/jira/browse/STORM-2913" title="STORM-2844 made autocommit and at-most-once storm-kafka-client spouts log warnings on every emit, because those modes don&#39;t commit the right metadata to Kafka" class="issue-link" data-issue-key="STORM-2913"><del>STORM-2913</del></a> properly. If we decide later that enable.auto.commit is unnecessary, we can remove NONE as an option.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> I am currently with limited Internet access. I will get my input on this by tomorrow. Thanks.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a> Thanks. I have most of the code changes ready for this and 2913, they're fairly small. I'll put up a pair of PRs soon. Happy to change them if they turn out to not be how we want to solve these issues.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>I feel <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a>'s issue is unlikely to having auto-commit or not, since there's only a slight difference between AT-MOST-ONCE and auto-commit as <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> explained. There might be little difference for the performance perspective since auto-commit is handled by Kafka hence they can make it as non-round-trip, but I guess it would be not a big deal.</p>

<p>I think investigating <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a>'s issue would be more helpful rather than supporting kind of workaround. The thing is why it was stuck, and why it couldn't be restored afterwards, and how does automatic backpressure play and how it affects the spout. What does "topology is stuck" mean? Even in back pressure state, non-spout should process tuples to reduce tuples in waiting queue so that back pressure flag can be turned off eventually. Topology shouldn't be stuck (except idle) at any chance.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>FYI to <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a>, the behavior of back pressure will be completely renewed in Storm 2.0.0.<br/>
<a href="https://issues.apache.org/jira/browse/STORM-2306" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-2306</a><br/>
Please refer the doc in the issue. I guess it makes more sense to you, but please comment to the issue if you find some concerns. The change is included in huge patch, so there's no chance to port back to 1.x version line.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> </p>

<p>I need to clarify a bit it seems.</p>

<p>The current proposal (the code currently in the PR) should not have a performance hit compared to the pre-1.2.0 code. It simply moves a bit of commit logic from the KafkaConsumer to the spout.</p>

<p>When we're talking about removing NONE, it is with the understanding that we'd want to benchmark it first so we can tell if using AT_MOST_ONCE instead of NONE has any real performance impact. I haven't benchmarked anything yet, so when I say that moving to AT_MOST_ONCE might incur a performance hit, I'm just speculating.</p>

<p>As Jungtaek mentioned, we're not going to remove NONE if it is a lot faster than AT_MOST_ONCE, but if it turns out there's no difference between the two (e.g. because the time spent sending commit requests to Kafka might be meaningless compared to time spent on other tasks), we would have no reason to keep NONE.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<blockquote>
<p>My belief is that our alerting topology gets stuck when there's a huge peak of messages rate because somehow we saturate Storm by exceeding its capacity, and/or maybe we hit some race condition that hands the spouts in a way that "normal load" would never trigger.</p></blockquote>

<p>I'd also add that what you're describing here is something you can likely avoid by configuring your topologies differently. For example, you can set topology.max.spout.pending to limit how many pending tuples will be allowed into the topology at a time, so a spike in messages written to Kafka doesn't trigger a flood of messages in Storm. Even without this configuration, the topology shouldn't hang (more likely an OOME would occur). I'd encourage you to try to get some more information (e.g. Storm UI screenshots, worker/nimbus/supervisor logs, thread dumps of the workers) if it happens again, and raise a new issue here so we can try to figure out what's happening.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen">avermeerbergen</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a></p>

<p>We do set topology.max.spout.pending, and we already had a discussion in the past that until Storm 2.x, we had to use our by-pass to overcome the limitations of Storm 1.x back-pressure:</p>

<p>See:  <a href="https://mail-archives.apache.org/mod_mbox/storm-user/201705.mbox/%3CCADeKz6qZG12mN-=Gf+mPTa1Jwxdk8_wWZ1NPbCgsLX7fgA4k9g@mail.gmail.com%3E" class="external-link" target="_blank" rel="nofollow noopener">https://mail-archives.apache.org/mod_mbox/storm-user/201705.mbox/%3CCADeKz6qZG12mN-=Gf+mPTa1Jwxdk8_wWZ1NPbCgsLX7fgA4k9g@mail.gmail.com%3E</a></p>

<p>(at subsequent posts). We observe no OOM, just the consumption from Spouts stops, regardless of whether it our own Kafka spout, the old Storm Kafka client or the newer Storm Kafka Client.</p>

<p>Back to Storm 1.2.0 : any idea when I could test a fix which removes the cycling "WARN" message about metadata ?</p>

<p>Thanks!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> I have a 1.x branch here <a href="https://github.com/srdo/storm/tree/STORM-2914-2913-1.x" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/srdo/storm/tree/STORM-2914-2913-1.x</a> with both 2914 and 2913 fixed. I'd appreciate if you would build and try out this fix.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> I am reviewing this patches today. I have going over the discussion since yesterday to put all the pieces together.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> the decision to keep or not keep NONE basically boils down to, do we want to support Kafka's enable.auto.commit or no? If we want to support this option, then the processing guarantee is technically NONE because a given record/tuple can be processed 0, 1, 2 or more times. In this case we cannot remove NONE.</p>

<p>That leads me to ask why are we removing the option to set the Kafka property enable.auto.commit and throwing an exception. This is a Kafka feature that is available to every Kafka consumer, so why should the KafkaSpout, which is technically a Kafka consumer, be any different?<br/>
 <br/>
If the goal is to avoid the WARN exception that was getting printed when enable.auto.commit=true, the obvious thing to do in my opinion is to simply not log the message if the processing guarantee is not AT_LEAST_ONCE.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a><br/>
What I suggest was to simplify the processing guarantee. Most of our supporting spouts guarantee either at-most-once or at-least-once, except exactly-once. So for me unless the gap between none and at-most-once is huge (like performance), I don't feel any benefit to have it and make KafkaSpout more complicated, not considering logic, but considering maintaining cost (mainly troubleshooting).</p>

<p>I think we are logging message as WARN because that's the thing users should be aware of. Based on my sense, simply not logging at all doesn't make sense. If it is ignorable, why it is logged as WARN? If it is the thing to handle, why we are going to swallow?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a> I am OK with supporting only at-most-once or at-least-once. Do you know how other frameworks deal with Kafka auto.commit mode? If they don't allow it altogether then Storm would be in an equal playing field and performance would not matter much at that point in my opinion. It will simply become a feature that is not supported.</p>

<p>As far as the WARN level, initially when I created the PR with that change I had it with level DEBUG and I never felt that it should be WARN to begin with. During reviews someone asked for it to be WARN and I did it, but never felt it should be WARN, and I still don't think it should be.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a><br/>
<a href="https://github.com/apache/flink/blob/1440e4febd589e320f846a2725e98aec8ee43e7f/docs/dev/connectors/kafka.md#kafka-consumers-offset-committing-behaviour-configuration" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/flink/blob/1440e4febd589e320f846a2725e98aec8ee43e7f/docs/dev/connectors/kafka.md#kafka-consumers-offset-committing-behaviour-configuration</a></p>

<p>Flink clearly documents how "enable.auto.commit" option plays with Flink's checkpoint mechanism. Key statement is below:</p>

<blockquote>
<p>"Note that the Flink Kafka Consumer does not rely on the committed offsets for fault tolerance guarantees. The committed offsets are only a means to expose the consumer's progress for monitoring purposes."</p></blockquote>

<p>Based on the statement, Flink doesn't have any concern about committing to Kafka, which is very different to Storm since storm-kafka-client is completely relying on Kafka committed offset.</p>

<p>Here's Spark doc for streaming Kafka source:<br/>
<a href="https://github.com/apache/spark/blob/55b8cfe6e6a6759d65bf219ff570fd6154197ec4/docs/streaming-kafka-0-10-integration.md#storing-offsets" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/spark/blob/55b8cfe6e6a6759d65bf219ff570fd6154197ec4/docs/streaming-kafka-0-10-integration.md#storing-offsets</a></p>

<p>Relevant statements are below:</p>

<blockquote>
<p>Kafka has an offset commit API that stores offsets in a special Kafka topic. By default, the new consumer will periodically auto-commit offsets. This is almost certainly not what you want, because messages successfully polled by the consumer may not yet have resulted in a Spark output operation, resulting in undefined semantics. This is why the stream example above sets "enable.auto.commit" to false.</p></blockquote>

<p>Spark community also avoids to couple with "enable.auto.commit". Checkpoint would be done with different way, and even guide regarding storing offsets to kafka itself, they don't recommend using that option.</p>

<p>cc. <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Whatever we decide, it may be better to document such guide so that we could let users to be aware of any effects when enabling such option.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a><br/>
 I think my reasoning for not using/allowing enable.auto.commit basically matches the reasoning for why we switched off of the subscribe API; It fits the spout poorly, it increases complexity for no reason, and we can do better because we're likely not the use case that setting was designed for. I'll add that there's no reason for us to use a feature in the consumer just because it's there, we should have a reason to want to support it.</p>

<p>My guess for why enable.auto.commit exists is because it is useful for simple processors that do something like</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">while</span> (<span class="code-keyword">true</span>) {
  records = kafkaConsumer.poll();
  processRecords(records);
}
</pre>
</div></div>
<p>The reason the setting is a great idea here, is that the (async) commit for the previous iteration's records happens during poll (assuming the auto.commit.interval.ms has expired, otherwise the commit may be postponed a bit), so you get nice at-least-once semantics "for free", without bothering with committing manually. You could get basically the same result by doing</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">while</span> (<span class="code-keyword">true</span>) {
  records = kafkaConsumer.poll();
  processRecords(records);
  <span class="code-keyword">if</span> (timer.isExpired()) {
    kafkaConsumer.commitAsync()
    timer.reset();
  }
}
</pre>
</div></div>
<p>but enabling auto commit by default provides a nice out of the box experience for the consumer where the user doesn't have to remember to commit.</p>

<p>If processRecords is asynchronous, you end up with a situation like ours where autocommit causes you to lose the ability to guarantee processing. The only reason we would want to keep it around is for people who really don't care about whether tuples are processed, because there may be a performance difference between doing</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">while</span> (<span class="code-keyword">true</span>) {
  records = kafkaConsumer.poll();
  processRecords(records);
  <span class="code-keyword">if</span> (timer.isExpired()) {
    kafkaConsumer.commitAsync()
    timer.reset();
  }
}
</pre>
</div></div>
<p>and doing at-most-once</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
<span class="code-keyword">while</span> (<span class="code-keyword">true</span>) {
  records = kafkaConsumer.poll();
  kafkaConsumer.commitSync();
  processRecords(records);
}
</pre>
</div></div>
<p>due to the greater number of blocking calls in the at-most-once case.</p>

<p>The reason I think we should remove enable.auto.commit support and provide our own NONE ("no guarantees, but probably the fastest option") implementation, is that <a href="https://issues.apache.org/jira/browse/STORM-2844" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-2844</a> slightly changed the semantics of EARLIEST and LATEST. For the at-least-once case EARLIEST now means "start over when the topology is redeployed", where it previously meant "start over when the worker is restarted". I think the new semantics are much more useful than the old ones. We can get this behavior because we commit metadata to Kafka as part of commits, which tells us which topology committed the offset.</p>

<p>We didn't update the at-most-once or NONE cases to store metadata, which is causing <a href="https://issues.apache.org/jira/browse/STORM-2913" title="STORM-2844 made autocommit and at-most-once storm-kafka-client spouts log warnings on every emit, because those modes don&#39;t commit the right metadata to Kafka" class="issue-link" data-issue-key="STORM-2913"><del>STORM-2913</del></a>. We can (and should) support the same behavior for EARLIEST and LATEST for at-most-once, and we can do it very easily by adding metadata to those commits.</p>

<p>For the NONE/enable.auto.commit case, we can't support the same behavior. It's not possible for us to specify metadata that the consumer should include in autocommits. This means that we'd have to special case EARLIEST/LATEST to specify that they do one thing for at-least-once/at-most-once and a completely different (and much less useful) thing for autocommit. We'd also have to add special case code to support it. This was the initial reason I wanted to get rid of autocommit, because it would add more complexity than I thought it was worth.</p>

<p>Since there isn't really anything complicated about implementing NONE on our own without enable.auto.commit (we even have the timer already), I think it's much more attractive to implement NONE ourselves, because it means we can make EARLIEST/LATEST behave consistently regardless of ProcessingGuarantee, rather than special casing autocommit out and creating a trap for users.</p>
<blockquote><p>If the goal is to avoid the WARN exception that was getting printed when enable.auto.commit=true, the obvious thing to do in my opinion is to simply not log the message if the processing guarantee is not AT_LEAST_ONCE.</p></blockquote>
<p>I hope I've explained now why I don't think it's good enough just to ignore the warning if we're in autocommit mode. We'd still be making EARLIEST/LATEST inconsistent across ProcessingGuarantees. Besides, if we just swallow the error we're still throwing an exception per emitted message which I don't think we want.</p>
<blockquote><p>As far as the WARN level, initially when I created the PR with that change I had it with level DEBUG and I never felt that it should be WARN to begin with. During reviews someone asked for it to be WARN and I did it, but never felt it should be WARN, and I still don't think it should be.</p></blockquote>
<p>Yes, that was me. I disagree. The log is in a place where we've caught an exception because we couldn't deserialize some data we expected to be there. It changes the spout's behavior a lot when it's hit during initialization as well (ignore the committed offset and do EARLIEST/LATEST if those strategies are in use), and ideally we never want to hit the error, so I really think DEBUG and INFO levels are too low. With this patch and 2913, the warning should only show up if it's the first time you start the spout on a partition and there were previous commits for that partition and consumer group (barring bugs).</p>
<blockquote><p>That leads me to ask why are we removing the option to set the Kafka property enable.auto.commit and throwing an exception</p></blockquote>
<p>The exception isn't new. It's there because I don't want people to set enable.auto.commit manually. The spout needs to know whether the consumer is autocommitting, because it needs to change its behavior. The best way to ensure that the consumer and spout settings are in sync is to make people set NONE, and using that setting to set enable.auto.commit. We're throwing an error if people try to set enable.auto.commit themselves in 2.x and in 1.x we're warning them that they should set the processing guarantee to NONE instead of using the property, then overwriting their settings.</p>

<p>All this being said, I haven't checked what the performance impact of at-most-once vs NONE is, but I suspect it would vary by latency to Kafka. If we want to drop NONE entirely, I would be okay with it, if we're sure that the performance penalty for using at-most-once for cases where you could also use NONE is not too great.</p>

<p>Sorry about the wall of text, hope this addresses your issues.</p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> you might also be interested to read this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a><br/>
As I showed the example of Flink and Spark and I believe others are doing it, we need to make data source being "adapted" to Storm's nature when implementing spout. We are making KafkaSpout being adapted to how Spout is working, and if we are going to support stateful exactly-once via checkpoint and need to make storm-kafka-client supporting this, we may also need to have a proper strategy like Spark and Flink, get rid of committing to kafka or at least not relying on that for checkpoint, since it is ideal to store checkpoint into "same" durable storage.</p>

<p>We need to take this into account while designing the spout and restrict some features if necessary, instead of letting users to use whatever they want to, but with the risks we don't guide whichever will be. If we really want to provide the feature which doesn't conform to Storm's nature, we should warn users, guiding the possible risks are best, but at least guiding that they should know what is exactly they're doing now.</p>

<p>That's why I'm not in favor of having NONE processing guarantee semantic only for storm-kafka-client, but if it really has enough value to take the risks (if there's huge gap between at-most-once and none in point of  performance view) we may need to provide the feature, with guiding the risks.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Btw, not a blocker but just food for thought is that the timing of commitment. We are committing in nextTuple() which would not be called in several states, flooding messages exceed max spout pending, in backpressure state, etc. Sometimes it can get somewhat longer even acks are coming (making timer being ineffective at that moment), and if there's crash, topology will discard acks completely while nextTuple() is not called and reprocess from start of backpressure at any luck or even worse, which might bring another backpressure and make loop if backpressure comes from the characteristic of records in poll.</p>

<p>May need to double check about the model of backpressure which <a href="https://issues.apache.org/jira/browse/STORM-2306" title="Redesign Messaging Subsystem, switch to JCTools Queues and introduce new Backpressure model" class="issue-link" data-issue-key="STORM-2306"><del>STORM-2306</del></a> will bring (I guess we could get rid of this situation with new backpressure model), but at least 1.x version line it will persist. I don't think handling commit in ack() with timer would work, so current approach would be our best effort in current situation (handling it from both ack()/nextTuple() might be an alternative solution to mitigate), but we may want to consider such case to improve spout's lifecycle and behavior in future (not only KafkaSpout but also Storm's spout).</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Correction: If new model of backpressure doesn't adjust a rate but just block calling nextTuple(), when backpressure bubbles up to spout it may encounter same situation. New model doesn't always block spout in backpressure state, so the chance to encounter will become much less.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen">avermeerbergen</a>:</i>
<p>Hello <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a></p>

<p>I have build storm-kafka-client.jar from your 1.x branch, tests are starting now.</p>

<p>While it's too early to say if it's OK or not, I noticed the following warning in one of our topologies's log:</p>

<p> </p>

<p>2018-02-01 13:08:25.016 o.a.k.c.u.AppInfoParser Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;INFO&#93;</span> Kafka version : 0.10.2.1 2018-02-01 13:08:25.016 o.a.k.c.u.AppInfoParser Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;INFO&#93;</span> Kafka commitId : e89bffd6b2eff799 2018-02-01 13:08:25.076 o.a.s.k.s.KafkaSpout Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;INFO&#93;</span> Partitions revoked. [consumer-group=StormPodsyncTopology_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ, consumer=org.apache.kafka.clients.consumer.KafkaConsumer@5cb084d9, topic-partitions=[]] 2018-02-01 13:08:25.077 o.a.s.k.s.KafkaSpout Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;INFO&#93;</span> Partitions reassignment. [task-ID=2, consumer-group=StormPodsyncTopology_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ, consumer=org.apache.kafka.clients.consumer.KafkaConsumer@5cb084d9, topic-partitions=<span class="error">&#91;podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-15, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-4, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-3, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-6, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-5, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-0, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-2, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-1, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-12, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-11, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-14, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-13, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-8, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-7, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-10, podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-9&#93;</span>] 2018-02-01 13:08:25.083 o.a.k.c.c.i.AbstractCoordinator Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;INFO&#93;</span> Discovered coordinator ec2-34-242-207-227.eu-west-1.compute.amazonaws.com:9092 (id: 2147483644 rack: null) for group StormPodsyncTopology_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ. 2018-02-01 13:08:25.116 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=63683, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-15, offset=63682, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.125 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=67701, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-4, offset=67700, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.130 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=69382, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-3, offset=69381, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.134 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=78943, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-6, offset=78942, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.135 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=76603, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-5, offset=76602, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.136 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=63190, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-0, offset=63189, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.137 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=73541, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-2, offset=73540, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.140 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=70191, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-1, offset=70190, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.142 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=70966, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-12, offset=70965, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.143 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=71945, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-11, offset=71944, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.148 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=74705, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-14, offset=74704, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.149 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=70042, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-13, offset=70041, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.153 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=70662, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-8, offset=70661, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.154 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=64628, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-7, offset=64627, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.155 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=72981, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-10, offset=72980, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.156 o.a.s.k.s.i.CommitMetadata Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;WARN&#93;</span> Failed to deserialize [OffsetAndMetadata{offset=61852, metadata='</p>
{topic-partition=podsync_RealTimeSupervision_zDHZGCMKTHGySBqVw9DiqQ-9, offset=61851, numFails=0, thread='Thread-4-podOrTenantsFromKafkaSpout-executor[2 2]'}
<p>'}]. Error likely occurred because the last commit for this topic-partition was done using an earlier version of Storm. Defaulting to behavior compatible with earlier version 2018-02-01 13:08:25.159 o.a.s.k.s.KafkaSpout Thread-4-podOrTenantsFromKafkaSpout-executor<span class="error">&#91;2 2&#93;</span> <span class="error">&#91;INFO&#93;</span> Initialization complete</p>

<p> </p>

<p>I don't know yet whether or not it'll be repeated forever, I'll check later.</p>

<p>Is this something expected from your perspective?</p>

<p>Could it be just a "transition" effect ?</p>

<p>Best regards,</p>

<p>Alexandre Vermeerbergen</p>

<p> </p>

<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen">avermeerbergen</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a></p>

<p>Also I noticed that in the artifacts built by Maven from your 1.x branch, there's this new JAR file:</p>

<p>storm-kafka-monitor-1.2.1-SNAPSHOT.jar</p>

<p>What is it meant for? Should I include it into my topologies "BigJars" ?</p>

<p>Worth mentioning: in Storm UI, the stats on Kafka consumption are no longer displayed.. is it related to this new JAR ?</p>

<p>Best regards,</p>

<p>Alexandre</p>

<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pshah">pshah</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> That jar is for displaying the Kafka spout offset lag in Storm UI if topology has a Kafka Spout. The jar should already be present in the toollib directory of your installation and you should not need to include it in your topologies' jar. Can you confirm that?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a></p>

<p>Yes, those logs are expected to occur, but only once per spout instance. Once the spout commits, the log should not be shown again.</p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a></p>

<p>Sure, if we're moving to a stateful checkpointing model we might want to use that mechanism's storage to commit Kafka offsets as well. I don't think there's a reason to switch off of Kafka storage yet though.</p>

<p>If I'm understanding you correctly, you're saying that it's fine to restrict how users can configure the spout and consumer, because some configurations don't make sense for Storm (right?) If so, I agree and I think it's what we're trying to do with e.g. the code around enable.auto.commit already (<a href="https://github.com/apache/storm/blob/master/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java#L461)," class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/blob/master/external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java#L461),</a> or by not offering an option to use the subscribe API. This issue has largely the same kind of intent.</p>

<p>I'm hoping to have a bit of time to compare NONE and AT_MOST_ONCE this weekend. I'm wondering if <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> has any interest in comparing these two settings as well, since you have a real workload?</p>

<p>Regarding the spout only committing when nextTuple is called, you have a point. I'm not sure why we couldn't move commit to happen during acks for the at-least-once case (the only one where acks matter). I don't have any insight into whether the case you describe would be likely to happen, but I'd think that it would often be possible to "luck into" a commit during the tuple sequence that leads up to triggering backpressure.</p>

<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen">avermeerbergen</a>:</i>
<p>Hello <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pshah" class="user-hover" rel="pshah">Parag Shah</a> and <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a>,</p>

<p>Regarding storm-kafka-monitoring*.jar : thanks for your explanation, it was indeed already in our toollib/ directory from Storm 1.2.0RC2:</p>

<p><tt><span class="error">&#91;root@ip-172-31-18-84 storm-stable&#93;</span># ls toollib/</tt><br/>
 <tt>storm-kafka-monitor-1.2.0.jar  storm-kafka-monitor-1.2.0-javadoc.jar  storm-kafka-monitor-1.2.0-sources.jar</tt></p>

<p>But since I was no longer seeing Kafka spout stats in Nimbus UI's log, and it seem that this was broken because of some missing class issue (very strange that it's trying to load a class from *javadoc.jar" file, isn't it ?) :</p>

<p><tt>org.apache.storm.utils.ShellUtils$ExitCodeException: Error: Could not find or load main class .usr.local.Storm.storm-stable.toollib.storm-kafka-monitor-1.2.0-javadoc.jar</tt></p>

<p><tt>        at org.apache.storm.utils.ShellUtils.runCommand(ShellUtils.java:231) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.utils.ShellUtils.run(ShellUtils.java:161) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.utils.ShellUtils$ShellCommandExecutor.execute(ShellUtils.java:371) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.utils.ShellUtils.execCommand(ShellUtils.java:461) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.utils.ShellUtils.execCommand(ShellUtils.java:444) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.utils.TopologySpoutLag.getLagResultForKafka(TopologySpoutLag.java:163) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.utils.TopologySpoutLag.getLagResultForNewKafkaSpout(TopologySpoutLag.java:189) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.utils.TopologySpoutLag.lag(TopologySpoutLag.java:57) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.ui.core$topology_lag.invoke(core.clj:805) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.ui.core$fn__9572.invoke(core.clj:1165) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.compojure.core$make_route$fn__5965.invoke(core.clj:100) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.compojure.core$if_route$fn__5953.invoke(core.clj:46) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.compojure.core$if_method$fn__5946.invoke(core.clj:31) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.compojure.core$routing$fn__5971.invoke(core.clj:113) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at clojure.core$some.invoke(core.clj:2570) ~<span class="error">&#91;clojure-1.7.0.jar:?&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.compojure.core$routing.doInvoke(core.clj:113) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at clojure.lang.RestFn.applyTo(RestFn.java:139) ~<span class="error">&#91;clojure-1.7.0.jar:?&#93;</span></tt><br/>
 <tt>        at clojure.core$apply.invoke(core.clj:632) ~<span class="error">&#91;clojure-1.7.0.jar:?&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.compojure.core$routes$fn__5975.invoke(core.clj:118) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.cors$wrap_cors$fn__8880.invoke(cors.clj:149) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.json$wrap_json_params$fn__8827.invoke(json.clj:56) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.multipart_params$wrap_multipart_params$fn__6607.invoke(multipart_params.clj:118) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.reload$wrap_reload$fn__7890.invoke(reload.clj:22) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.ui.helpers$requests_middleware$fn__6860.invoke(helpers.clj:52) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.ui.core$catch_errors$fn__9747.invoke(core.clj:1428) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.keyword_params$wrap_keyword_params$fn__6527.invoke(keyword_params.clj:35) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.nested_params$wrap_nested_params$fn__6570.invoke(nested_params.clj:84) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.params$wrap_params$fn__6499.invoke(params.clj:64) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.multipart_params$wrap_multipart_params$fn__6607.invoke(multipart_params.clj:118) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.flash$wrap_flash$fn__6822.invoke(flash.clj:35) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.middleware.session$wrap_session$fn__6808.invoke(session.clj:98) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.util.servlet$make_service_method$fn__6357.invoke(servlet.clj:127) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.util.servlet$servlet$fn__6361.invoke(servlet.clj:136) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.ring.util.servlet.proxy$javax.servlet.http.HttpServlet$ff19274a.service(Unknown Source) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt><br/>
 <tt>        at org.apache.storm.shade.org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:654) ~<span class="error">&#91;storm-core-1.2.0.jar:1.2.0&#93;</span></tt></p>

<p>Now in the .jar files built using Stig's 1.x branch, I see that there are two different versions of storm-kafka-monitor Jar file:</p>

<p><tt>./storm-<a href="https://issues.apache.org/jira/browse/STORM-2914" title="Remove enable.auto.commit support from storm-kafka-client" class="issue-link" data-issue-key="STORM-2914"><del>STORM-2914</del></a>-2913-1.x/external/storm-kafka-monitor/target/storm-kafka-monitor-1.2.1-SNAPSHOT.jar</tt><br/>
 <tt>./storm-<a href="https://issues.apache.org/jira/browse/STORM-2914" title="Remove enable.auto.commit support from storm-kafka-client" class="issue-link" data-issue-key="STORM-2914"><del>STORM-2914</del></a>-2913-1.x/external/storm-kafka-monitor/target/original-storm-kafka-monitor-1.2.1-SNAPSHOT.jar</tt></p>

<p>which one would be the "right" one? what's the difference between them?</p>

<p><ins>Note:</ins> I couldn't resist so I tried removing everything from toollib/, then I copied <tt>storm-kafka-monitor-1.2.1-SNAPSHOT.jar</tt> into it and then I restarted everything (including Nimbus UI) => the metrics on Kafka spout are back in Nimbus UI <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>  but why was it broken by default in Storm 1.2.0 RC2 ?</p>

<p> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a>:</p>

<p>The WARN messages seems now gone, so from my perspective it confirms that <a href="https://issues.apache.org/jira/browse/STORM-2913" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-2913</a> is solved in your 1.x branch, thanks!</p>

<p>Regarding your proposal to compare compare NONE and AT_MOST_ONCE with my load: why not, but then please give me clear instructions on how to setup my Kafka spout, so I won't do any mistake in this setup (and I guess some output will be printed to confirm the setup).</p>

<p>Best regards,</p>

<p>Alexandre Vermeerbergen</p>

<p> </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a></p>

<p>You've got my points. Thanks for reading my comment thoughtfully.</p>

<p>Regarding commit in ack(), we may need to take into consideration that unlike nextTuple() it is called only when spout receives ack. It should be easiest one but we don't want to commit every ack (acks may not come sequentially, and performance matters). But if we apply timer to ack(), we need to deal with edge case: not yet timed-out but received some acks afterwards which are "last" acks. Possible to deal with, but might be tricky. That's why I mentioned hybrid approach - allowing commit to Kafka in both nextTuple() and ack(), which still has such edge case as well but the chance is much smaller.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pshah">pshah</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> Not sure how the javadoc and sources jar got in to the binary distribution. On 1.1.1 binary distribution package i only see </p>

<p>storm-kafka-monitor-1.1.1.jar</p>

<p>That should be okay and not create any issue in loading spout offset lags in storm ui. I will verify in the latest RCs as well. Can you double check as well to see if the other jars are present and if that is creating an issue in displaying the offset lags?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a>, <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a>, <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen" class="user-hover" rel="avermeerbergen">Alexandre Vermeerbergen</a> </p>

<p>Let’s try to quickly agree on the specific action points that will help us move past any blockers and release Storm 1.2.0. I suggest the following, which will give us a good balance between making this release stable, but at the same time will give us the flexibility to accommodate changes that make sense in the future.</p>

<ol>
	<li>(Optional) Rename ProcessingGuarantee.NONE to ProcessingGuarantee.NO_GUARANTEES or UNDEFINED (as Spark uses). I think that either of these names are clearer and emphasize that this option is likely not what the users wants</li>
	<li>Annotate the enum ProcessingGuarantee with <b>@Unstable</b> and document that ProcessingGuarantee.NO_GUARANTEES may be removed in the near future.</li>
	<li>For now we keep supporting NO_GUARANTEES mode, but there is a good chance that we will remove it down the line, unless for performance reasons. We will do a benchmark for this.</li>
	<li>NO_GUARANTEES will use <b>kafkaConsumer.commitAsync(offsetsToCommit, null);</b> as <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> proposed in the PR.</li>
	<li>Document that ProcessingGuarantee.NO_GUARANTEES has a behavior identical to auto.commit.enable=true but disregards this property altogether (if auto.commit.enable is set the KafkaSpoutConfig will throw an exception). Furthermore, the commits will not be based on a timer as controlled by auto.commit.interval.ms, but rather using the following strategy (<a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> perhaps you can add the details in here).</li>
	<li>AT_MOST_ONCE will use <b>kafkaConsumer.commitSync(offsetsToCommit);</b> as <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> proposed in the PR</li>
	<li>AT_LEAST_ONCE will stay as is (in the current code and in the PR) because this option requires no change.</li>
</ol>


<p>Slightly unrelated with this PR but to address <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a>’s comment on how to persist the offsets. The next versions of the KafkaSpout (1.2.x, 1.3+.x and 2.x) should have offsets persistence pluggable and abstracted behind an interface. The first implementation of this interface will commit offsets to Kafka as it is currently doing. In the future we can add support to write offsets to arbitrary durable storage.</p>

<p>If we agree on this, <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> can you please incorporate these changes in the PRs you created and target having them merged by end of Monday. Thanks everyone.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=avermeerbergen">avermeerbergen</a>:</i>
<p>Hello <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a></p>

<p>+1 for your proposal to get things decided quickly !</p>

<p>In your 7 questions lists, my vote is:</p>

<p>+1 for item #5 : because I want to trust you guys on the meaning of "identical" in this sentence: "<em>Document that ProcessingGuarantee.NO_GUARANTEES has a behavior identical to auto.commit.enable=true but disregards this property altogether (if auto.commit.enable is set the KafkaSpoutConfig will throw an exception)</em>".</p>

<p>And I have no problem changing the way I configure Kafka spout to get this "identical" behavior if documentation is crystal clear about it. The documentation needs to be very clear about how to replace legacy use of auto.commit.enable and related setting by newer ones.</p>

<p>+ 0 for other items, meaning I am neutral there, as I'm selfishly concerned by my need to keep a behavior "identical" to auto.commit.enable=true.</p>

<p>I just hope that "identical" means "no noticeable difference in performance".</p>

<p>Best regards,</p>

<p>Alexandre Vermeerbergen</p>

<p>PS: I also hope that in next RC, the toollib/ directory will be cleaned up from <b>src</b> and <b>javadoc</b> artifacts, which seems to break the display of Kafka spout statistics in Nimbus UI in 1.2.0 RC2 (see my previous comment about this in this thread).**</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=srdo">srdo</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a></p>

<p>I had time to do some tests. I've added a storm-perf topoloy that uses storm-kafka-client instead of storm-kafka, here <a href="https://github.com/apache/storm/pull/2545." class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/2545.</a></p>

<p>I ran the topology from the mentioned PR for 10 minutes, with NONE and AT_MOST_ONCE against a Kafka instance with small messages preloaded ("testMessage"). Because the commitSync call in AT_MOST_ONCE is blocking and thus likely affected by latency to Kafka, I tried running the topology both against a local Kafka node, and against one set up in an Azure VM.</p>

<p>I've attached the raw results.  <span class="nobr"><a href="https://issues.apache.org/jira/secure/attachment/12909104/12909104_storm-kafka-modes.ods" title="storm-kafka-modes.ods attached to STORM-2914">storm-kafka-modes.ods<sup><img class="rendericon" src="https://issues.apache.org/jira/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span> . Here's the averages for spout acks/s for each test:</p>
<ul>
	<li>NONE local: 124778.6</li>
	<li>AT_MOST_ONCE local: 120226.6</li>
</ul>


<ul>
	<li>NONE Azure (25-40 ms latency): 34593.6</li>
	<li>AT_MOST_ONCE Azure: 10374.8</li>
</ul>


<p>As you can see it's not too expensive to use AT_MOST_ONCE locally instead of NONE, but once latency is introduced, it looks much cheaper to use NONE if possible. I'm a little surprised the difference is so large, I would have expected something more like 50% throughput for AT_MOST_ONCE compared to NONE due to doubling the number of blocking calls in nextTuple when polling.</p>

<p>I'm not too experienced with benchmarking, so it's possible that I've made a mistake somewhere, or this benchmark may be too synthetic to really show anything.</p>

<p>For now I think we should keep NONE. I'll do as Hugo suggests and rename it and mark it unstable. I agree with Hugo's list, but we should consider waiting to hide committing/loading commits behind an interface until we actually need to store commits outside Kafka. Once we know what a second implementation needs, it'll probably be easier to make a good generic interface, rather than one that only fits our current use case, and I also would prefer not to add pluggability if we end up not needing it. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro">hmclouro</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> Thanks for working on this over the weekend. I am going to start reviewing your patch now. We are in agreement on the best timing to create an interface to support pluggable offsets storage. I also don't thing we should do it now, i.e. for 1.2.0, hence why I hinted at 1.2.x, 1.3+.x and 2.x, basically some release after 1.2.0. Thanks.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a>, merged into master.</p>

<p>Since there's some diverge between master and 1.x-branch for storm-kafka-client, I'd like to wait for <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Srdo" class="user-hover" rel="srdo">Stig Rohde Døssing</a> to submit pull request against 1.x branch. We can merge it afterwards.</p>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-2913">STORM-2913</a>]</i>

**Comment by jira-importer** [comment]
<i>[Epic: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-2710%20label%3Aepic">STORM-2710</a>]</i>
